### PR TITLE
fix: infer correct hook handler

### PR DESCRIPTION
**Pull request by Uzlopak**
This PR fixes the issue, that you could define async hook handlers, but they could contain the done callback. 

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
Thanks! I think this should fix quite a lot of problem for folks.

File: test/types/hooks.test-d.ts [hunk 1]
```diff
@@ -392,3 +398,20 @@ server.addHook('preClose', function (done) {
...
+
+server.addHook('onRequest', async function (request, reply) {
+  expectType<FastifyRequest>(request)
+  expectType<FastifyReply>(reply)
+  expectType<FastifyInstance>(this)
+})
+
+expectError(server.addHook('onRequest', async function (request, reply, done) {
+  expectType<FastifyInstance>(this)
+}))
```
Comment (Uzlopak) [inline test/types/hooks.test-d.ts:417]:
```suggestion
expectError(server.addHook('onRequest', async function (request, reply, done) {}))
```

**Comment by Uzlopak** [comment]
Changing to Draft, because it still needs some love

**Comment by Uzlopak** [comment]
Ok, I thought about it and wanted to inform you about one drawback:

Because the method headers of the hook handlers are all different, typescript will throw the error on the wrong place.

image

It is an improvement, because we get an error in typescript when async and callback is mixed. But you get errors like   `Argument of type '"onClose"' is not assignable to parameter of type '"preClose"'`

File: types/instance.d.ts [hunk 2]
```diff
@@ -30,6 +30,8 @@ export interface PrintRoutesOptions {
   includeHooks?: boolean
 }
 
+type AsyncFunction = (...args: any) => Promise<any>;
```
Comment (metcoder95) [inline types/instance.d.ts:33]: Should we also support `PromiseLike<any>`?

**Review by metcoder95 — COMMENTED** [review]
Nice 🚀 

Haven't played in deep with the Types from fastify, but something like this could be ported to the `route` definition?

File: types/instance.d.ts (hunk 2 above)
Comment (Uzlopak) [reply types/instance.d.ts:33]: Well, we define the functions as async. So they should be Promise and not promiselike. But i am open for it. Maybe fix this first than we can think about optimizing it.

**Review by metcoder95 — APPROVED** [review]

**Comment by Uzlopak** [comment]
Regarding the issue of wrong error... I will check if sorting the hooks by descending arity of the handler can help or not .

**Comment by Uzlopak** [comment]
much better
@metcoder95 

image

**Review by metcoder95 — APPROVED** [review]
LGTM 👍

**Review by mcollina — CHANGES_REQUESTED** [review]
On a second look, this is not correct. the `done` callback must not be used when the function is async

**Comment by Uzlopak** [comment]
@mcollina 

It throws typings errors, when you define the function async and provide a callback parameter.
You can actually see this in my last screenshot. Also in the typings test, all of the combinations are expectError

**Review by mcollina — APPROVED** [review]
lgtm

Sorry I misunderstood this

**Pull request merged**